File: app/src/main/java/com/faz/nothing/MainActivity.java
package com.faz.nothing;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import java.util.Set;

public final class MainActivity extends AppCompatActivity implements LocationService.LocationSetUpdateListener {

    private Button startBtn;
    private Button stopBtn;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        startBtn = findViewById(R.id.start);
        stopBtn = findViewById(R.id.stop);

        Intent serviceIntent = new Intent(this, LocationService.class);
        startBtn.setOnClickListener(v -> startService(serviceIntent));
        stopBtn.setOnClickListener(v -> stopService(serviceIntent));
    }

    @Override
    protected void onResume() {
        super.onResume();
        drawLocations();
        LocationService.locationUpdateListeners.add(this);
    }

    @Override
    protected void onPause() {
        super.onPause();
        LocationService.locationUpdateListeners.remove(this);
    }

    @Override
    public void onLocationSetChanged() {
        drawLocations();
    }

    private void drawLocations() {
        Set<Location> locations = LocationService.locationSet;
        Logger.log("Drawing " + locations);
    }
}

File: app/src/main/java/com/faz/nothing/MainActivity.kt
package com.faz.nothing

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button

class MainActivity : AppCompatActivity(), LocationService.LocationSetUpdateListener {

    private val startBtn: Button by lazy { findViewById<Button>(R.id.start) }
    private val stopBtn: Button by lazy { findViewById<Button>(R.id.stop) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val serviceIntent = Intent(this, LocationService::class.java)
        startBtn.setOnClickListener {
            startService(serviceIntent)
        }
        stopBtn.setOnClickListener {
            stopService(serviceIntent)
        }
    }

    override fun onResume() {
        super.onResume()
        drawLocations()
        LocationService.locationUpdateListeners.add(this)
    }

    override fun onPause() {
        super.onPause()
        LocationService.locationUpdateListeners.remove(this)
    }

    override fun onLocationSetChanged() {
        drawLocations()
    }

    private fun drawLocations() {
        val locations = LocationService.locationSet
        Logger.log("Drawing $locations")
    }
}